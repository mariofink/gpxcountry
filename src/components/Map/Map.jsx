import { useEffect, useRef } from "react";
import "leaflet/dist/leaflet.css";
import { map, tileLayer, marker, LayerGroup, divIcon, polyline } from "leaflet";

const Map = ({ lat, lng, onClick, waypoints }) => {
  const osMap = useRef(null);
  const markersLayer = useRef(null);
  const lineLayer = useRef(null);
  // init map
  useEffect(() => {
    if (osMap.current !== null) return; // if map has already been initialised
    // Setting tap to false to fix Safari issues (see: https://github.com/Leaflet/Leaflet/issues/7255)
    osMap.current = map("leafletmap", { tap: false }).setView([lat, lng], 13);
    markersLayer.current = new LayerGroup();
    markersLayer.current.addTo(osMap.current);
    lineLayer.current = new LayerGroup();
    lineLayer.current.addTo(osMap.current);
    tileLayer("https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png", {
      attribution:
        '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a>',
      useCache: true,
    }).addTo(osMap.current);
    osMap.current.on("click", (e) => {
      onClick(e.latlng);
    });
  }, [lat, lng, onClick]);
  // update map
  useEffect(() => {
    markersLayer.current.clearLayers();
    lineLayer.current.clearLayers();
    waypoints.map((waypoint, index) => {
      const markerIcon = divIcon({
        className: "map-marker text-white bg-gray-700",
        iconSize: [20, 20],
        html: `<span>${index + 1}</span>`,
      });
      const mapMarker = marker([waypoint.lat, waypoint.lng], {
        icon: markerIcon,
      });
      markersLayer.current.addLayer(mapMarker);
    });
    if (waypoints.length > 1) {
      const line = polyline(waypoints, { color: "#1086e8", weight: 6 });
      lineLayer.current.addLayer(line);
    }
  }, [waypoints]);
  return <div id="leafletmap" className="h-full"></div>;
};

export default Map;
