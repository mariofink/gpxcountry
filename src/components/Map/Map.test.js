import React from "react";
import renderer from "react-test-renderer";
import Map from "./Map";

describe("Map", () => {
  it("renders", () => {
    const tree = renderer
      .create(
        <Map lat="50.5238" lng="10.0204" onClick={() => {}} waypoints={[]} />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
