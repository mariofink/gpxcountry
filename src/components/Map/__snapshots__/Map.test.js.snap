// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map renders 1`] = `
<div
  className="h-full"
  id="leafletmap"
/>
`;
